### ISSUE TITLE
MRS arguments appear to be dropped in the LaViDa LLaMA inference path

### ISSUE THREAD

**Issue by zichuanJI** [hong-sujung/MPD-PAC#2]
Hi, thanks for releasing the paper and code. While reproducing the LaViDa-L ablation, I observed that the MRS arguments appear to be dropped on the ordinary LLaMA inference path. Could you clarify the intended evaluation entry point or configuration?

Tested commit: `7ce18852cb9af779cf64c00ea7f8d272414b296c`.

The reproduction uses `jacklishufan/lavida-llada-v1.0-instruct`, BF16, `block_type=llama`, `block_group_size=1`, `use_cache=False`, and activation checkpointing disabled. The author's `llada_ours/modeling_llada.py` and `llada_ours/generate.py` are byte-identical to the pinned commit; no MRS fix is applied. The evaluation adapter only selects the model class, loads BF16/local files, and decodes the generated suffix.

### Observed call path

1. [`LLaDAModel.forward`, ordinary per-block call](https://github.com/hong-sujung/MPD-PAC/blob/7ce18852cb9af779cf64c00ea7f8d272414b296c/LaViDa/llava/model/language_model/llada_ours/modeling_llada.py#L1538) does not forward `rope`, `mode`, `slope`, or `center` to the block.
2. [`LLaDALlamaBlock.forward`, ordinary attention call](https://github.com/hong-sujung/MPD-PAC/blob/7ce18852cb9af779cf64c00ea7f8d272414b296c/LaViDa/llava/model/language_model/llada_ours/modeling_llada.py#L1089) also omits these parameters.
3. [`RotaryEmbedding.forward`](https://github.com/hong-sujung/MPD-PAC/blob/7ce18852cb9af779cf64c00ea7f8d272414b296c/LaViDa/llava/model/language_model/llada_ours/modeling_llada.py#L482) converts `rope=None` to zero strength.

With `rope=0.1, mode="sigmoid", slope=12, center=0.6` supplied at the model entry point, a forward pre-hook on each rotary module observes the omitted arguments. A small random-weight CPU model also produces bitwise-identical logits when toggling `rope` between 0 and 0.1. This small-model check tests argument propagation, not benchmark quality.

### Minimal reproduction

Run this in the repository's LaViDa environment (with its LaViDa directory on `PYTHONPATH`). This uses a tiny random-weight CPU model and does not download model weights or alter model code.

```python
"""Run inside the author's LaViDa environment; does not change model code."""
import json
import torch
from llava.model.language_model.llada_ours.modeling_llada import LLaDAModel
from llava.model.language_model.llada_ours.configuration_llada import (
    ModelConfig, BlockType, LayerNormType, ActivationType,
)

cfg = ModelConfig(
    d_model=64, n_heads=4, n_layers=2, mlp_hidden_size=128,
    block_type=BlockType.llama, layer_norm_type=LayerNormType.rms,
    activation_type=ActivationType.silu, rope=True, max_sequence_length=96,
    vocab_size=128, embedding_size=128, attention_dropout=0,
    embedding_dropout=0, residual_dropout=0, init_device='cpu',
    eos_token_id=125, pad_token_id=126, mask_token_id=127,
)
torch.manual_seed(42)
model = LLaDAModel(cfg).eval()
model.set_activation_checkpointing(None)
ids = torch.randint(0, 125, (1, 64))
observed = []


def record(module, args, kwargs):
    observed.append({'rope': args[2], 'mode': args[3],
                     'slope': args[4], 'center': kwargs.get('center')})


handles = [b.rotary_emb.register_forward_pre_hook(record, with_kwargs=True)
           for b in model.transformer.blocks]
kw = dict(mode='sigmoid', slope=12, center=0.6)
with torch.inference_mode():
    a = model(ids, rope=0.0, **kw).logits
    b = model(ids, rope=0.1, **kw).logits
for handle in handles:
    handle.remove()
print(json.dumps({'bitwise_equal': torch.equal(a, b),
                  'max_abs_logit_difference': (a-b).abs().max().item(),
                  'rotary_arguments': observed}, indent=2))
```

Observed output (startup messages omitted):

```json
{
  "bitwise_equal": true,
  "max_abs_logit_difference": 0.0,
  "rotary_arguments": [
    {
      "rope": null,
      "mode": "linear",
      "slope": null,
      "center": null
    },
    {
      "rope": null,
      "mode": "linear",
      "slope": null,
      "center": null
    },
    {
      "rope": null,
      "mode": "linear",
      "slope": null,
      "center": null
    },
    {
      "rope": null,
      "mode": "linear",
      "slope": null,
      "center": null
    }
  ]
}
```

### Independent two-call forwarding check

For a separate control, I forwarded `rope`, `mode`, `slope`, `center`, and `attn` through just these two ordinary calls. I kept the author's sampler byte-identical and made no rotary-cache or zero-prior shortcut changes. The comparison is limited to the ordinary `block_type=llama`, `block_group_size=1`, non-checkpointed inference path; it is not a proposed fix for every other configuration.

```diff
--- a/llava/model/language_model/llada_ours/modeling_llada.py
+++ b/llava/model/language_model/llada_ours/modeling_llada.py
@@ -1086,7 +1086,7 @@
                 self.attention, q, k, v, attention_bias, layer_past=layer_past, use_cache=use_cache,block_mask=block_mask
             )
         else:
-            att, cache = self.attention(q, k, v, attention_bias, layer_past=layer_past, use_cache=use_cache,block_mask=block_mask)
+            att, cache = self.attention(q, k, v, attention_bias, layer_past=layer_past, use_cache=use_cache, block_mask=block_mask, rope=rope, mode=mode, slope=slope, center=center, attn=attn)
 
         # Unpack result based on whether attn is requested
         # if attn:
@@ -1535,7 +1535,7 @@
                     )
                 else:
                     # shape: (batch_size, seq_len, d_model)
-                    x, cache = block(x, attention_bias=attention_bias, layer_past=layer_past, use_cache=use_cache,block_mask=block_mask)
+                    x, cache = block(x, attention_bias=attention_bias, layer_past=layer_past, use_cache=use_cache, block_mask=block_mask, rope=rope, mode=mode, slope=slope, center=center, attn=attn)
                 if attn_key_values is not None:
                     assert cache is not None
                     attn_key_values.append(cache)
```

On the tiny CPU check, the patched zero-strength output remained bitwise identical to the original, while enabling MRS produced a maximum absolute logit change of `3.606081008911133e-06`. Hooks observed the requested `rope` values and `mode="sigmoid", slope=12` in the patched run.

### Completed LaViDa-L comparison

Both implementations completed Baseline / +MPS / +MRS / +Both on all four evaluation sets, with locally verified result backups. They used the same BF16 model weights, examples, prompts, seed 42, and generation settings. MPS used `prior=0.3`; MRS used `rope=0.1` (the implementation additionally applies `low_strength=0.1`), with `mode="sigmoid", slope=12, center=0.6, k=3`.

The generation `(steps, length)` pairs were MME `(2,2)`, DetailCaps `(128,256)`, LLaVA-Bench `(64,128)`, and Ferret `(48,96)`, with one block of that length, `schedule="shift"`, shift 3, temperature 0, `remasking="low_confidence"`, and `prefix_lm=False`. These choices are reproduction settings, not a claim that all unspecified paper settings were recovered.

In the **unmodified implementation**, Baseline and +MRS had identical prediction text for every example; +MPS and +Both also matched on every example. Across the unmodified and patched implementations, Baseline and +MPS matched for every example. The counts below compare each patched variant against its corresponding unmodified variant:

| Evaluation set | Examples per variant | Baseline changed | +MPS changed | +MRS changed | +Both changed |
|---|---:|---:|---:|---:|---:|
| MME (full) | 2374 | 0 | 0 | 18 | 20 |
| DetailCaps (fixed subset) | 100 | 0 | 0 | 26 | 95 |
| LLaVA-Bench In-the-Wild (full) | 60 | 0 | 0 | 37 | 41 |
| Ferret (fixed subset) | 100 | 0 | 0 | 44 | 52 |

DetailCaps and Ferret use fixed 100-example subsets whose IDs may differ from the paper. LLaVA-Bench and Ferret were independently scored with DeepSeek Flash as an alternative judge; identical predictions sometimes received different judge scores. CAPTURE also showed small scoring-stage differences between independent runs on identical control predictions. I therefore base the propagation observation on hooks and exact prediction equality, not on judge-score differences or a claim of improved benchmark quality.

Is a different evaluation entry point/configuration intended for MRS, or should these arguments be forwarded through both calls? I am reporting the behavior of this public inference path without drawing conclusions about the paper's experimental results.
